### PR TITLE
Fixes for ublox SARA modem driver

### LINKED ISSUE #21328: Apparent network context leak with offloading driver (u-blox Sara r4)
**Describe the bug**
I have an issue when I try to make multiple MQTT connection/disconnection.
I focused on net_context management.
Indeed in net_context_get function, I saw that I reached CONFIG_NET_MAX_CONTEXTS (equal to 10 in my prj.conf).
For your information, I use UBLOX sara r4 driver (on zephyr 1.14.1) and I wonder if contexts are well managed in this revision (or new one?), especially when socket is closed ?

**To Reproduce**
Compiling https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/net/mqtt_publisher:
1. mkdir build; cd build
2. cmake -DBOARD=boards\nrf52840_pca10059 -DSHIELD=sparkfun_sara_r4 -DCONF_FILE="prj.conf overlay-wncm14a2a.conf" (I am not sure of this configuration because I did this with our own board)
3. make

**Expected behavior**
After 10 (equal to CONFIG_NET_MAX_CONTEXTS) connection/disconnection in TCP, you should have this behavior: 
[try_to_connect:233] mqtt_connect: -2 <ERROR>.

**Impact**
It is not possible to communicate anymore in MQTT (TCP) because it is not possible to connect to broker after CONFIG_NET_MAX_CONTEXTS attemps

**Screenshots or console output**
image

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain ARM-GCC
 - Tag: v1.14.1

**Additional context**
I also saw an issue in UBLOX SARA R4 socket close management when URC +UUCLOSE happens.
I am not sure that this issue is linked to our issue. Moreover, maybe it is well managed in new revision

### LINKED ISSUE #21764: [SARA-R4] MQTT publisher not working - Impossible to connect to broker
**Describe the bug**
I see an issue when I try to use MQTT Publisher through SARA R4 modem driver. Indeed, it is impossible to conenct to MQTT server/broker.
I checked that Network connection was ok.

**To Reproduce**
Compiling https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/net/mqtt_publisher::
1. mkdir build; cd build
2. cmake -DBOARD=boards\nrf52840_pca10059 -DSHIELD=sparkfun_sara_r4 -DCONF_FILE="prj.conf overlay-wncm14a2a.conf" (I am not sure of this configuration because I did this with our own board)
3. make

**Expected behavior**
When tryin to connect to MQTT broker, I always have following issue:
MQTT connect failed -111
So it is impossible to communicate.

**Impact**
No MQTT communication on my board which is protocol chosen for our need by our customer.

**Screenshots or console output**
image

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: ARM-GCC
 - Tag: v2.0

**Additional context**
I have follwoing traces when I snif UART bus
AT+USOCR=6
+USOCR: 0

OK
AT+USOCO=0,"83.166.154.87",1883
OK
AT+USOWR=0,30
@
+USOWR: 0,30

OK
AT+USOWR=0,30
@
+USOWR: 0,30

OK
AT+USOWR=0,30
@
+USOWR: 0,30

OK

+UUSORD: 0,4
AT+USOWR=0,30
@
+CME ERROR: 3

### PR DESCRIPTION
**Pull request by mike-scott**
This PR is a combination of fixes for a wide variety of issues noticed on the SARA-R4 modem:
- MQTT wasn't working correctly
- http_get sample was crashing
- echo_client wasn't working
- odd behavior was happening in LwM2M client under load

Some additional authors contributed patches which were added here for easy testing due to the # of changes:
@weinholtendian and @hwilmers

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/21764
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/21328

### CODE REVIEW

**Comment by mike-scott** [comment]
These fixes are related to: https://github.com/zephyrproject-rtos/zephyr/issues/21764  But don't quite fix all issues.  MQTT is still having problems due to the way it uses recv() with perfectly sized buffers to parse the socket data.

The SARA-R4 modem AT commands may have a bug where the full packet data needs to be queried at a time.

**Comment by mike-scott** [comment]
@gpaquet85 @weinholtendian @WilliamGFish FYI

**Review by brooksprumo — APPROVED** [review]
I was previously running into an assertion fault because a network interface pointer was NULL and being dereferenced, when using Particle Boron. This PR fixed that crash. Yay!

**Comment by mike-scott** [comment]
NOTE: More work coming on this PR.

**Comment by jukkar** [comment]
> MQTT is still having problems due to the way it uses recv() with perfectly sized buffers to parse the socket data.

@mike-scott What kind of problems are there in MQTT, can you elaborate more?
cc: @rlubos

**Comment by mike-scott** [comment]
> > MQTT is still having problems due to the way it uses recv() with perfectly sized buffers to parse the socket data.
> 
> @mike-scott What kind of problems are there in MQTT, can you elaborate more?
> cc: @rlubos

@jukkar @rlubos 
Yes, sorry to be a bit vague.  I was trying to make better use of the modem's data buffer by only pulling data as it was needed via the recv() call (based on the length of the buffer passed in).

The MQTT library parses individual packets bytes at a time, so this leads to successive calls to the modem to retrieve data for the same packet.  The way this modem driver works is it keeps track of how much data in in the modem's buffer (split up by packet). My handling of the packet sizes didn't allow for partial retrieval of packets in this way.

I need to think about how to handle this.

**Comment by hwilmers** [comment]
Nice fixes, this works here with a Sara R410.

**Comment by mike-scott** [comment]
@gpaquet85 @WilliamGFish @weinholtendian @hwilmers
Can you please retest with the full set of patches?

NOTE: To test MQTT, you will also need this PR: https://github.com/zephyrproject-rtos/zephyr/pull/22249

**Comment by mike-scott** [comment]
Here is the way I tested 3 of the samples for the SparkFun SARA-R4 shield (on a K64F):

LwM2M client (connecting to the IP of http://leshan.eclipse.org/):
```
west build -b frdm_k64f -s zephyr/samples/net/lwm2m_client -- \
  -DSHIELD=sparkfun_sara_r4 \
  -DCONFIG_NET_CONFIG_PEER_IPV4_ADDR=\"23.97.187.154\" \
  -DCONFIG_NET_IPV6=n -DCONFIG_NET_CONFIG_NEED_IPV6=n \
  -DCONFIG_MODEM_UBLOX_SARA_R4_APN=\"<APN>\" \
  -DCONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO=\"<manual MCCMNO>\" \
  -DCONFIG_COAP_INIT_ACK_TIMEOUT_MS=10000 -DCONFIG_MODEM_SHELL=y \
  -DCONFIG_LWM2M_SERVER_DEFAULT_PMIN=1 \
  -DCONFIG_LWM2M_SERVER_DEFAULT_PMAX=360 \
  -DCONFIG_LOG_STRDUP_MAX_STRING=128 \
  -DCONFIG_LOG_STRDUP_BUF_COUNT=10
```

HTTP GET sample:
```
west build -b frdm_k64f -s zephyr/samples/net/sockets/http_get -- \
  -DSHIELD=sparkfun_sara_r4 \
  -DCONFIG_MODEM_UBLOX_SARA_R4_APN=\"<APN>\" \
  -DCONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO=\"<manual MCCMNO>\" \
  -DCONFIG_MODEM_SHELL=y \
  -DCONFIG_LOG_IMMEDIATE=n \
  -DCONFIG_NET_CONFIG_NEED_IPV4=y
```

MQTT Client (also need the changes here: https://github.com/zephyrproject-rtos/zephyr/pull/22249)
```
west build -b frdm_k64f -s zephyr/samples/net/mqtt_publisher -- \
  -DSHIELD=sparkfun_sara_r4 \
  -DCONFIG_NET_CONFIG_PEER_IPV4_ADDR=\"5.196.95.208\" \
  -DCONFIG_NET_IPV6=n -DCONFIG_NET_CONFIG_NEED_IPV6=n \
  -DCONFIG_MODEM_UBLOX_SARA_R4_APN=\"<APN>\" \
  -DCONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO=\"<manual MCCMNO>\" \
  -DCONFIG_MODEM_SHELL=y \
  -DCONFIG_NET_CONFIG_NEED_IPV4=y
```

**Comment by WilliamGFish** [comment]
> Here is the way I tested 3 of the samples for the SparkFun SARA-R4 shield (on a K64F):
> 
Now I see why you have been avoiding NET_OFFLOAD, I am using the Particle boards which use the U2, R4 and esp32 none of which have network port. This is why I have been suggesting to continue with the network offloading standard.

Maybe the nice guys at Nordic will send you a couple of Dev Kits to experiment with as the SparkFun shields should be compatible.

**Comment by mike-scott** [comment]
> Now I see why you have been avoiding NET_OFFLOAD, I am using the Particle boards which use the U2, R4 and esp32 none of which have network port. This is why I have been suggesting to continue with the network offloading standard.
> 
> Maybe the nice guys at Nordic will send you a couple of Dev Kits to experiment with as the SparkFun shields should be compatible.

I have all 3 Particle.IO boards: Xenon, Argon and Boron.  I don't have a VZW SIM for use on my Boron so it makes things difficult, but I can certainly help with the Argon ESP8622 bring up (I'm looking at the PR: https://github.com/zephyrproject-rtos/zephyr/pull/21492)

Other than TLS support and possibly DNS, what functionality were you looking for on the NET_OFFLOAD APIs?

Looking at esp8622, it has internal methods for handling SSL-based connections which would need to use the setsockopt offloaded APIs (AT+CIPSSLCONF).

**Comment by WilliamGFish** [comment]
> I have all 3 Particle.IO boards: Xenon, Argon and Boron. I don't have a VZW SIM for use on my Boron so it makes things difficult, but I can certainly help with the Argon ESP8622 bring up (I'm looking at the PR: #21492)

Be warned you will need to change the firmware on the esp32 before you can use it as the particle firmware is "locked down" so will give you odd results.

**Comment by mike-scott** [comment]
> Be warned you will need to change the firmware on the esp32 before you can use it as the particle firmware is "locked down" so will give you odd results.

Was there a specific version to use from this list?
https://www.espressif.com/en/support/download/at

**Comment by WilliamGFish** [comment]
> > Be warned you will need to change the firmware on the esp32 before you can use it as the particle firmware is "locked down" so will give you odd results.
> 
> Was there a specific version to use from this list?
> https://www.espressif.com/en/support/download/at

You may find this useful: https://learn.adafruit.com/adding-a-wifi-co-processor-to-circuitpython-esp8266-esp32/program-particle-argon

**Comment by tsvehagen** [comment]
> > Now I see why you have been avoiding NET_OFFLOAD, I am using the Particle boards which use the U2, R4 and esp32 none of which have network port. This is why I have been suggesting to continue with the network offloading standard.
> > Maybe the nice guys at Nordic will send you a couple of Dev Kits to experiment with as the SparkFun shields should be compatible.
> 
> I have all 3 Particle.IO boards: Xenon, Argon and Boron. I don't have a VZW SIM for use on my Boron so it makes things difficult, but I can certainly help with the Argon ESP8622 bring up (I'm looking at the PR: #21492)
> 
> Other than TLS support and possibly DNS, what functionality were you looking for on the NET_OFFLOAD APIs?
> 
> Looking at esp8622, it has internal methods for handling SSL-based connections which would need to use the setsockopt offloaded APIs (AT+CIPSSLCONF).

The main reason for me _not_ wanting to offload SSL in the esp8266, or any external device really, is that data will be passed in clear text over the UART.

**Comment by WilliamGFish** [comment]
> The main reason for me _not_ wanting to offload SSL in the esp8266, or any external device really, is that data will be passed in clear text over the UART.

I too am using a plain text buffer to feed the offloaded network i/f via the UART

**Comment by hwilmers** [comment]
I tested this PR with Sara R410, with following results:
- MQTT client works fine
- echo_client (TCP) works fine with small chunks (up to 1023 bytes)
- echo_client (TCP) fails with chunks >= 1024 bytes

I then discovered that in case of these failures, the +USOWR fails with an error message from the modem ("ERROR"), immediately followed by a +UUSORD message from the modem.

To me it looks like a solution could be to let the application poll for incoming data in between multiple calls to send(). So should this be solved in the driver, or rather in the application?

**Comment by hwilmers** [comment]
Apropos TLS: Is there a way to use this driver with TLS, other than offloading it? Are there any hooks one can use to use mbedTLS?

**Comment by WilliamGFish** [comment]
> Apropos TLS: Is there a way to use this driver with TLS, other than offloading it? Are there any hooks one can use to use mbedTLS?

There are options to use mbedTLS in zephyr (CONFIG_MBEDTLS_BUILTIN)...it might but something to look at.

**Comment by mike-scott** [comment]
> I tested this PR with Sara R410, with following results:
> 
> * MQTT client works fine
> * echo_client (TCP) works fine with small chunks (up to 1023 bytes)
> * echo_client (TCP) fails with chunks >= 1024 bytes
> 
> I then discovered that in case of these failures, the +USOWR fails with an error message from the modem ("ERROR"), immediately followed by a +UUSORD message from the modem.
> 
> To me it looks like a solution could be to let the application poll for incoming data in between multiple calls to send(). So should this be solved in the driver, or rather in the application?

Thanks for testing, let me take a look at the >= 1024 use case to see where the fix would best be located.  In a perfect world, we fix that in the driver so that samples don't need to be changed.  However, there is some "bad behavior" in the samples sometimes which can be changed.

**Comment by mike-scott** [comment]
> The main reason for me _not_ wanting to offload SSL in the esp8266, or any external device really, is that data will be passed in clear text over the UART.

When I looked at how the SSL certificates would be setup for ESP8266, it seems like they needed to be written to flash to avoid this situation.

`AT+CIPSSLCCONF=2` means (taken from AT command doc):
```
bit1:if set to be 1, CA will be enabled, so ESP8266 can verify SSL server; if 0, then will not.
```

And the example AT command uses are:
```
AT+CIPSNTPCFG=1,8 // set SNTP timezone
AT+CIPSNTPTIME? // get SNTP time
AT+CIPSSLCCONF=2
AT+CIPSTART="SSL","192.168.3.38",8443
```

There doesn't seem to be a time when cleartext SSL information would be used over UART.

**Comment by tsvehagen** [comment]
> There doesn't seem to be a time when cleartext SSL information would be used over UART.

Right but then when you send or read data over that connection the data will be sent unencrypted over the UART? Maybe I'm missing something 🤔

**Comment by mike-scott** [comment]
> Right but then when you send or read data over that connection the data will be sent unencrypted over the UART? Maybe I'm missing something

Yes, this would be true.  UART lines are a fairly easy physical attack vector.
Unfortunately, this isn't limited to SSL connections.  The Wifi AP password is sent in clear text, etc.

**Comment by mike-scott** [comment]
Updated with 2 patches to fixup the return values for socket offload and vtable functions.  They weren't standard POSIX values.

**Comment by gpaquet85** [comment]
@mike-scott really thanks for your support on this case.
I will made some tries next week and let you know.
I had also a look on MQTT PR. I will integrate it to know if it is ok from my side.

**Comment by WilliamGFish** [comment]
> Updated with 2 patches to fixup the return values for socket offload and vtable functions. They weren't standard POSIX values.

Some sterling work @mike-scott , I'll pull this PR and do some testing at the beginning of next week.

**Comment by mike-scott** [comment]
I have a few more patches to get on top of this PR, but it needs to be rebased now with the socket offload rework being merged.  The TOPIC-GPIO branch will also cause conflicts once it's merged.

**Comment by gpaquet85** [comment]
@mike-scott I made some tries. For the moment basic MQTT publish is ok. Thanks for this. However I still have some fails when I run MQTT publish loop. I investigate right now on this and let you know.

**Comment by mike-scott** [comment]
Rebased on today's master with the last set of changes:
- Remove hex mode.  It limits the send/receive MTU on SARA-R4 to 512 which is too restrictive.  We can now handle this inside the modem commands by returning EAGAIN if the command doesn't have enough data to finish parsing.
- parse_params() shouldn't parse as many delimiters as it can, instead it should honor the arg_count value and stop once it receives that # of params
- And a few minor changes

**Comment by mike-scott** [comment]
> I tested this PR with Sara R410, with following results:
> 
> * MQTT client works fine
> * echo_client (TCP) works fine with small chunks (up to 1023 bytes)
> * echo_client (TCP) fails with chunks >= 1024 bytes
> 
> I then discovered that in case of these failures, the +USOWR fails with an error message from the modem ("ERROR"), immediately followed by a +UUSORD message from the modem.
> 
> To me it looks like a solution could be to let the application poll for incoming data in between multiple calls to send(). So should this be solved in the driver, or rather in the application?

I have a separate PR to fix echo_client for net interfaces where the MTU is < 1280: https://github.com/zephyrproject-rtos/zephyr/pull/22448

**Comment by mike-scott** [comment]
> run MQTT publish loop. I invest

Please retry with the latest version of this PR.  I had a few more fixes to land on top today.  Sorry about that.

**Comment by gpaquet85** [comment]
> 
> 
> > @mike-scott I made some tries. For the moment basic MQTT publish is ok. Thanks for this. However I still have some fails when I run MQTT publish loop. I investigate right now on this and let you know.
> 
> Please retry with the latest version of this PR. I had a few more fixes to land on top today. Sorry about that.

@mike-scott I made some tries. It works. Really thanks for your support on this case :)

**Comment by rlubos** [comment]
@mike-scott Just out of curiosity, given the offloading rework got merged sooner than I expected, you had to rebase on top of this work. Does the new approach work fine for you, did you encounter any issues with these changes? (apart from extra work with conflict solving I guess)

**Comment by mike-scott** [comment]
Rebased on today's master (post TOPIC-GPIO merge):
- added patches to fix build issues due to GPIO changes

**Comment by mike-scott** [comment]
@jukkar Does this PR have any impact on the new GSM driver?  I don't think it does..

**Comment by jukkar** [comment]
> @mike-scott  Does this PR have any impact on the new GSM driver? I don't think it does..

Looks quite clean to me so there should be no issues with GSM driver.

**Comment by jukkar** [comment]
@mike-scott Some of these features would be useful to have in GSM driver too if the modem is a SARA modem, for example "support for network indication on SARA modems". Currently we need to add this to GSM modem init function, but I wonder if we could isolate these generic and common settings to some library or similar, which could be used from both the GSM driver and offloading driver.

File: drivers/modem/Kconfig.ublox-sara-r4
Comment (jukkar) [inline drivers/modem/Kconfig.ublox-sara-r4]: Why this is a string and not an int?

File: drivers/modem/modem_cmd_handler.c
Comment (jukkar) [inline drivers/modem/modem_cmd_handler.c]: Not really an issue but it would be nice to the reader if () would be used here, so `if ((data->match_buf_len - 1) < match_len) {`

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c]: Could we call `modem_atoi()` here so that we can handle error cases properly.

**Review by jukkar — CHANGES_REQUESTED** [review]

**Comment by mike-scott** [comment]
Need to rebase this again (I'll remove the last 2 patches with fixes that have been merged).   And address @jukkar comments.

**Comment by mike-scott** [comment]
> @mike-scott Just out of curiosity, given the offloading rework got merged sooner than I expected, you had to rebase on top of this work. Does the new approach work fine for you, did you encounter any issues with these changes? (apart from extra work with conflict solving I guess)

No, so far so good.  @rlubos thanks for the rework :)

File: drivers/modem/Kconfig.ublox-sara-r4
Comment (mike-scott) [reply drivers/modem/Kconfig.ublox-sara-r4]: Woops.  I picked up this patch from @weinholtendian -- I will make this an int.

File: drivers/modem/modem_cmd_handler.c
Comment (mike-scott) [reply drivers/modem/modem_cmd_handler.c]: I'll add parens.

File: drivers/modem/ublox-sara-r4.c
Comment (mike-scott) [reply drivers/modem/ublox-sara-r4.c]: Hah, yes.  Good catch.  I have the define already and didn't use it here.

File: drivers/modem/ublox-sara-r4.c
Comment (mike-scott) [reply drivers/modem/ublox-sara-r4.c]: Sounds good.  I'll rework this.

File: drivers/modem/ublox-sara-r4.c
Comment (mike-scott) [reply drivers/modem/ublox-sara-r4.c]: Yep.  Will change this.

**Comment by mike-scott** [comment]
- Rebased again on today's master
- Addressed @jukkar comments

NOTE: I also shortened `MODEM_UBLOX_SARA_R4_NETWORK_STATUS` config to:
`MODEM_UBLOX_SARA_R4_NET_STATUS`.

To help with review, here's the diff from the previous version:
```
diff --git a/drivers/modem/Kconfig.ublox-sara-r4 b/drivers/modem/Kconfig.ublox-sara-r4
index 17736e89eb..c6a373d6eb 100644
--- a/drivers/modem/Kconfig.ublox-sara-r4
+++ b/drivers/modem/Kconfig.ublox-sara-r4
@@ -77,21 +77,21 @@ config MODEM_UBLOX_SARA_R4_INIT_PRIORITY
 	  Note that the priority needs to be lower than the net stack
 	  so that it can start before the networking sub-system.
 
-config MODEM_UBLOX_SARA_R4_NETWORK_STATUS
+config MODEM_UBLOX_SARA_R4_NET_STATUS
 	bool "Enable support for network status indication"
 	help
 	  Choose this setting to use a modem GPIO pin as network indication.
 
-if MODEM_UBLOX_SARA_R4_NETWORK_STATUS
+if MODEM_UBLOX_SARA_R4_NET_STATUS
 
-config MODEM_UBLOX_SARA_R4_NETWORK_STATUS_PIN
-	string "Network status indication gpio_id"
-	default "16"
+config MODEM_UBLOX_SARA_R4_NET_STATUS_PIN
+	int "Network status indication GPIO ID"
+	default 16
 	help
 	  This setting is used to configure one of the modem's GPIO pins
 	  as a network status indication.  See the manual for the gpio ids
 	  and how they map to pin numbers.
 
-endif # MODEM_UBLOX_SARA_R4_NETWORK_STATUS
+endif # MODEM_UBLOX_SARA_R4_NET_STATUS
 
 endif # MODEM_UBLOX_SARA
diff --git a/drivers/modem/modem_cmd_handler.c b/drivers/modem/modem_cmd_handler.c
index 4d795d54ee..fcab704510 100644
--- a/drivers/modem/modem_cmd_handler.c
+++ b/drivers/modem/modem_cmd_handler.c
@@ -279,7 +279,7 @@ static void cmd_handler_process(struct modem_cmd_handler *cmd_handler,
 		match_len = net_buf_linearize(data->match_buf,
 					      data->match_buf_len - 1,
 					      data->rx_buf, 0, len);
-		if (data->match_buf_len - 1 < match_len) {
+		if ((data->match_buf_len - 1) < match_len) {
 			LOG_ERR("Match buffer size (%zu) is too small for "
 				"incoming command size: %u!  Truncating!",
 				data->match_buf_len - 1, match_len);
diff --git a/drivers/modem/ublox-sara-r4.c b/drivers/modem/ublox-sara-r4.c
index 5241bc018d..1bb846b778 100644
--- a/drivers/modem/ublox-sara-r4.c
+++ b/drivers/modem/ublox-sara-r4.c
@@ -206,11 +206,11 @@ static ssize_t send_socket_data(struct modem_socket *sock,
 	}
 
 	/*
-	 * Binary and ASCII mode allows sending 1024 bytes to the socket
-	 * in one command
+	 * Binary and ASCII mode allows sending MDM_MAX_DATA_LENGTH bytes to
+	 * the socket in one command
 	 */
-	if (buf_len > 1024) {
-		buf_len = 1024;
+	if (buf_len > MDM_MAX_DATA_LENGTH) {
+		buf_len = MDM_MAX_DATA_LENGTH;
 	}
 
 	/* The number of bytes written will be reported by the modem */
@@ -729,11 +729,11 @@ static void modem_reset(void)
 		/* use external SIM */
 		SETUP_CMD_NOHANDLE("AT+UGPIOC=23,0,0"),
 #endif
-#if defined(CONFIG_MODEM_UBLOX_SARA_R4_NETWORK_STATUS_PIN)
+#if defined(CONFIG_MODEM_UBLOX_SARA_R4_NET_STATUS_PIN)
 		/* enable the network status indication */
 		SETUP_CMD_NOHANDLE("AT+UGPIOC="
-				   CONFIG_MODEM_UBLOX_SARA_R4_NETWORK_STATUS_PIN
-				   ",2"),
+			STRINGIFY(CONFIG_MODEM_UBLOX_SARA_R4_NET_STATUS_PIN)
+			",2"),
 #endif
 		/* UNC messages for registration */
 		SETUP_CMD_NOHANDLE("AT+CREG=1"),
@@ -1130,11 +1130,11 @@ static ssize_t offload_recvfrom(void *obj, void *buf, size_t len,
 	}
 
 	/*
-	 * Binary and ASCII mode allows sending 1024 bytes to the socket
-	 * in one command
+	 * Binary and ASCII mode allows sending MDM_MAX_DATA_LENGTH bytes to
+	 * the socket in one command
 	 */
-	if (next_packet_size > 1024) {
-		next_packet_size = 1024;
+	if (next_packet_size > MDM_MAX_DATA_LENGTH) {
+		next_packet_size = MDM_MAX_DATA_LENGTH;
 	}
 
 	snprintk(sendbuf, sizeof(sendbuf), "AT+USO%s=%d,%d",
@@ -1288,10 +1288,8 @@ static int offload_getaddrinfo(const char *node, const char *service,
 	struct modem_cmd cmd = MODEM_CMD("+UDNSRN: ", on_cmd_dns, 1U, ",");
 	u32_t port = 0U;
 	int ret;
-	/* AT+UDNSRN=#,"[+128]"\r+NUL == 17 + 128 == 145
-	 * (allows for domain name of up to 128 characters)
-	 */
-	char sendbuf[145];
+	/* DNS command + 128 bytes for domain name parameter */
+	char sendbuf[sizeof("AT+UDNSRN=#,'[]'\r") + 128];
 
 	/* init result */
 	(void)memset(&result, 0, sizeof(result));
@@ -1318,7 +1316,7 @@ static int offload_getaddrinfo(const char *node, const char *service,
 	}
 
 	if (service) {
-		port = strtol(service, NULL, 10);
+		port = ATOI(service, 0U, "port");
 		if (port < 1 || port > USHRT_MAX) {
 			return EAI_SERVICE;
 		}
```

**Review by jukkar — APPROVED** [review]
LGTM

File: drivers/modem/ublox-sara-r4.c
Comment (gpaquet85) [inline drivers/modem/ublox-sara-r4.c]: what do you think about adding timeout test in order to be sure to exit loop ?
I had one case where I was always blocked in this loop.

**Review by erwango — APPROVED** [review]
Approved (shield-wise)

File: drivers/modem/ublox-sara-r4.c
Comment (mike-scott) [reply drivers/modem/ublox-sara-r4.c]: @gpaquet85 Can we hold off on that until after this is merged?  The loop doesn't change here, other than the values are checked at the top of the loop.  We can address it in a new PR with discussion.

**Comment by mike-scott** [comment]
@gpaquet85 @hwilmers @WilliamGFish @weinholtendian Please weigh in as to whether this is ready to be merged.   I'd like those that are interested to have a voice.

**Comment by WilliamGFish** [comment]
I'll have a look next week. Billy..On 7 Feb 2020 20:15, Michael Scott <notifications@github.com> wrote:@gpaquet85 @hwilmers @WilliamGFish @weinholtendian Please weigh in as to whether this is ready to be merged.   I'd like those that are interested to have a voice.

**Comment by mike-scott** [comment]
@jukkar I'm hoping this PR is classified as a "bugfix" and is still acceptable for v2.2?

**Review by gpaquet85 — APPROVED** [review]

File: drivers/modem/ublox-sara-r4.c
Comment (gpaquet85) [reply drivers/modem/ublox-sara-r4.c]: @mike-scott ok no problem. Indeed it can be adressed later

**Review by hwilmers — APPROVED** [review]

**Comment by WilliamGFish** [comment]
Was all looking good until I tried to use the U2 version then an odd **_CME ERROR: 133 | Requested service option not subscribed_**  

Turns out a typo which I think came from a cut-n-paste from an order version(_ublox-sara-r4.c_):
```
#if defined(CONFIG_MODEM_UBLOX_SARA_U2)
		/* set the APN */
		SETUP_CMD_NOHANDLE("AT+UPSD=0,1,\""
				CONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO "\""),

```
Should be:
```
#if defined(CONFIG_MODEM_UBLOX_SARA_U2)
		/* set the APN */
		SETUP_CMD_NOHANDLE("AT+UPSD=0,1,\""
				CONFIG_MODEM_UBLOX_SARA_R4_APN "\""),

```

**Comment by mike-scott** [comment]
@WilliamGFish Ok, easy fix.  Let me put a PR up.  Can you open an issue for that?  Zephyr is in bug-fix only mode now.

**Comment by WilliamGFish** [comment]
> @WilliamGFish Ok, easy fix. Let me put a PR up. Can you open an issue for that? Zephyr is in bug-fix only mode now.

Issue raised
driver: modem: sara-u2 error when connecting #22689

**Comment by Brietoe** [comment]
I've a question, pardon my ignorance if I'm overlooking something here.

I've implemented this driver and have it working for the most part. When I open a socket and send data to an echo server, I get back a response which shows as an unsolicited message with a 'number bytes ready' count. When I use the recv() function to read incoming data, I'm unable to get data that arrived before I called the recv() function but has already been handled with the modem_socket_data_ready() function. It looks like, while in modem_socket_wait_data() it only waits for fresh data to come in and doesn't detect recently buffered data.

Is this the proper functionality of recv() or am I doing something wrong?

**Comment by weinholtendian** [comment]
> Is this the proper functionality of recv() or am I doing something wrong?

@Brietoe TCP data should never disappear merely because recv() was called too late. The driver is in control of when it reads from the modem's socket buffer, and it does not need to read if it doesn't have anywhere to save the data. Sounds to me like a pretty severe bug if data disappears this way. UDP is a slightly different story.

**Comment by Brietoe** [comment]
@weinholtendian Okay, that's what I thought. I don't want to jump to conclusions and claim that the offload_recv() does indeed suffer from this issue because I had to make some modifications to the code to fit my use case. The data is still accessible through workarounds but I'm just not able to retrieve it using recv().

I'd love to know if anyone else is also experiencing this issue.

**Comment by WilliamGFish** [comment]
> Other than TLS support and possibly DNS, what functionality were you looking for on the NET_OFFLOAD APIs?

Hey @mike-scott 
Looking at the particle boards and TLS is now a big question. I'm considering getting a PPP set-up option for Boron if I can.

Billy..

**Pull request merged**
